### ISSUE TITLE
Allow hiding errors that applying a concept failed

### ISSUE THREAD

**Issue by olivergierke** [buschmais/jqa-core-framework#7]
With the advent of generic rules (e.g. in the Spring plugin), it's nothing unusual that applying a concept actually yields not types being affected by a concept. These failed applications usually don't cause any issues as subsequent rules based on those concepts will just not yield any results and thus don't fail.

An analysis report will still the concepts not being applicable as failures which creates quite a bit of noise in the report: first, they show "errors" which indicate something wrong with the code, when there isn't anything wrong as long as it's only the concepts. Second, they might mask actual violations as you just skim over the report and think: "Yeah, it's just concept applications."

Here's a sample output that ideally wouldn't report any issues. Of course, that hiding could be an opt-in thing via the Maven plugin or the like.

```
[INFO] --- jqassistant-maven-plugin:1.1.4:analyze (default) @ videoshop ---
[INFO] Executing analysis for 'videoshop'.
[INFO] Reading rules from directory /Users/olivergierke/Documents/workspace/tu-dresden/videoshop/jqassistant
[INFO] Executing group 'spring-boot:Strict'
[INFO] Executing group 'spring-component:Strict'
[INFO] Executing group 'spring-component:Default'
[INFO] Applying concept 'spring-component:Configuration' with severity: 'MINOR'.
[INFO] Applying concept 'spring-data:AnnotatedRepository' with severity: 'MINOR'.
[INFO] Applying concept 'spring-data:ImplementedRepository' with severity: 'MINOR'.
[INFO] Applying concept 'spring-component:Repository' with severity: 'MINOR'.
[INFO] Applying concept 'spring-component:Controller' with severity: 'MINOR'.
[INFO] Applying concept 'spring-component:Service' with severity: 'MINOR'.
[INFO] Applying concept 'spring-component:Component' with severity: 'MINOR'.
[INFO] Applying concept 'spring-component:VirtualDependency' with severity: 'MINOR'.
[INFO] Validating constraint 'spring-component:ControllerMustOnlyDependOnServicesOrRepositories' with severity: 'INFO'.
[INFO] Validating constraint 'spring-component:ServiceMustOnlyDependOnServicesOrRepositories' with severity: 'INFO'.
[INFO] Validating constraint 'spring-component:RepositoryMustOnlyDependOnRepositories' with severity: 'INFO'.
[INFO] Validating constraint 'spring-component:ControllerMustDependEitherOnServicesOrRepositories' with severity: 'INFO'.
[INFO] Executing group 'spring-boot:Default'
[INFO] Executing group 'spring-transaction:Default'
[INFO] Applying concept 'spring-transaction:TransactionalClass' with severity: 'MINOR'.
[INFO] Applying concept 'spring-transaction:TransactionalMethod' with severity: 'MINOR'.
[INFO] Validating constraint 'spring-transaction:TransactionalMethodMustNotBeInvokedFromSameClass' with severity: 'INFO'.
[INFO] Executing group 'spring-injection:Default'
[INFO] Validating constraint 'spring-injection:AvoidInitializingBean' with severity: 'INFO'.
[INFO] Validating constraint 'spring-injection:AvoidDisposableBean' with severity: 'INFO'.
[INFO] Applying concept 'spring-injection:BeanProducer' with severity: 'MINOR'.
[INFO] Applying concept 'spring-injection:Injectable' with severity: 'MINOR'.
[INFO] Validating constraint 'spring-injection:InjectablesMustNotBeInstantiated' with severity: 'INFO'.
[INFO] Validating constraint 'spring-injection:BeanProducerMustBeDeclaredInConfigurationComponent' with severity: 'INFO'.
[INFO] Validating constraint 'spring-injection:AvoidAwareInterfacesInFavorOfInjection' with severity: 'INFO'.
[INFO] Validating constraint 'spring-injection:FieldsInitializedByConstructorMustNotBeManipulated' with severity: 'INFO'.
[INFO] Applying concept 'spring-boot:Application' with severity: 'MINOR'.
[INFO] Validating constraint 'spring-boot:AllTypesInApplicationPackage' with severity: 'INFO'.
[INFO] Executing group 'spring-injection:Strict'
[INFO] Applying concept 'spring-injection:InjectionPoint' with severity: 'MINOR'.
[INFO] Validating constraint 'spring-injection:FieldInjectionIsNotAllowed' with severity: 'INFO'.
[ERROR] --[ Concept Application Failure ]----------------------------------
[ERROR] Concept: spring-boot:Application
[ERROR] Severity: MINOR
[ERROR] Labels all Spring Boot Applications with "Spring" and "Application".
[ERROR] -------------------------------------------------------------------
[ERROR] 

[ERROR] --[ Concept Application Failure ]----------------------------------
[ERROR] Concept: spring-component:Service
[ERROR] Severity: MINOR
[ERROR] Labels all types annotated with "@org.springframework.stereotype.Service" with "Spring", "Service"
[ERROR]             and "Component".
[ERROR]         
[ERROR] -------------------------------------------------------------------
[ERROR] 

[ERROR] --[ Concept Application Failure ]----------------------------------
[ERROR] Concept: spring-component:VirtualDependency
[ERROR] Severity: MINOR
[ERROR] Creates a DEPENDS_ON relation between Spring Components which are connected through interfaces or
[ERROR]             abstract types and adds a "virtual" property.
[ERROR]         
[ERROR] -------------------------------------------------------------------
[ERROR] 

[ERROR] --[ Concept Application Failure ]----------------------------------
[ERROR] Concept: spring-data:AnnotatedRepository
[ERROR] Severity: MINOR
[ERROR] Labels all types annotated with "@org.springframework.stereotype.Repository" with "Spring",
[ERROR]             "Repository" and "Component".
[ERROR]         
[ERROR] -------------------------------------------------------------------
[ERROR] 

[ERROR] --[ Concept Application Failure ]----------------------------------
[ERROR] Concept: spring-injection:BeanProducer
[ERROR] Severity: MINOR
[ERROR] Labels methods which are annotated with "@org.springframework.context.annotation.Bean" with
[ERROR]             "BeanProducer" and their return types with "Spring" and "Injectable".
[ERROR]         
[ERROR] -------------------------------------------------------------------
```

**Comment by olivergierke** [comment]
Oh, and one thing I found puzzling when I first encountered those: "Concept Application Failure" — What does that mean? Is my code breaking the application of the concept? It's just not yielding any results, right? Maybe this could be reported as "Applying concept didn't result in any project artifacts matching." or the like?

**Comment by nmervaillie** [comment]
I encountered the same confusing behaviour using classpath:Resolve rules (command line).
Sometimes the classpath:ResolveWrites rule doesn't do anything, especially with Spring based applications using dependency injection. 

```
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Executing group 'my-rules:Default'
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveType' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveParameterType' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveMember' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveInvokes' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveThrows' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveReturns' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveAnnotationType' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveWrites' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveImplements' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveFieldType' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveValue' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveReads' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveDependency' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:ResolveExtends' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Applying concept 'classpath:Resolve' with severity: 'MINOR'.
[main] INFO com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Verifying results, severity=CRITICAL
[main] WARN com.buschmais.jqassistant.scm.cli.task.AbstractAnalyzeTask - Concept failed: classpath:ResolveWrites, Severity: MINOR
```

**Comment by DirkMahler** [comment]
Should be solved by #9, i.e. severity thresholds for logging or failing a build and configurable default severities for rules.
